### PR TITLE
NXP-23410: Define/Implement default bulk get key/value pairs w/optims

### PR DESCRIPTION
**Pull request by ron1**
NXP-23410: Define/Implement default bulk get key/value pairs w/optimizations for MongoDB and Redis

### CODE REVIEW

File: nuxeo-core/nuxeo-core-mongodb/src/main/java/org/nuxeo/ecm/core/mongodb/kv/MongoDBKeyValueStore.java [hunk 1]
```diff
@@ -181,6 +186,67 @@ protected Object getObject(String key) {
         return value;
     }
 
+    /**
+     * @since 9.10
+     */
+    @Override
+    public Map<String, byte[]> get(Set<String> keys) {
+        Objects.requireNonNull(keys);
+        Map<String, byte[]> kvMap = new HashMap<>(keys.size());
```
Comment (efge) [inline nuxeo-core/nuxeo-core-mongodb/src/main/java/org/nuxeo/ecm/core/mongodb/kv/MongoDBKeyValueStore.java]: I would just call this `map` (same in other methods)

File: nuxeo-core/nuxeo-core-mongodb/src/main/java/org/nuxeo/ecm/core/mongodb/kv/MongoDBKeyValueStore.java [hunk 2]
```diff
@@ -181,6 +186,67 @@ protected Object getObject(String key) {
...
+        Block<Document> block = doc -> {
+            String key = doc.getString(ID_KEY);
+            Object value = doc.get(VALUE_KEY);
+            byte[] bytesValue;
+            if (value instanceof String) {
+                bytesValue = ((String) value).getBytes(UTF_8);
+            } else if (value instanceof Binary) {
+                bytesValue = ((Binary) value).getData();
+            } else {
```
Comment (efge) [inline nuxeo-core/nuxeo-core-mongodb/src/main/java/org/nuxeo/ecm/core/mongodb/kv/MongoDBKeyValueStore.java]: What about missing values (i.e., `value == null`). Please add a unit test for that and treat that case.

File: nuxeo-core/nuxeo-core-mongodb/src/main/java/org/nuxeo/ecm/core/mongodb/kv/MongoDBKeyValueStore.java [hunk 3]
```diff
@@ -181,6 +186,67 @@ protected Object getObject(String key) {
...
+                throw new UnsupportedOperationException(value.getClass().getName());
```
Comment (efge) [inline nuxeo-core/nuxeo-core-mongodb/src/main/java/org/nuxeo/ecm/core/mongodb/kv/MongoDBKeyValueStore.java]: You should log the key as well.

File: nuxeo-core/nuxeo-core-mongodb/src/main/java/org/nuxeo/ecm/core/mongodb/kv/MongoDBKeyValueStore.java [hunk 4]
```diff
@@ -181,6 +186,67 @@ protected Object getObject(String key) {
...
+            } else if (value instanceof Binary) {
+                byte[] bytes = ((Binary) value).getData();
+                try {
+                    strValue = bytesToString(bytes);
+                } catch (CharacterCodingException e) {
+                    // fall through to throw
+                }
+            }
+            if (strValue == null) {
+                throw new IllegalArgumentException("Value is not a String for key: " + key);
```
Comment (efge) [inline nuxeo-core/nuxeo-core-mongodb/src/main/java/org/nuxeo/ecm/core/mongodb/kv/MongoDBKeyValueStore.java]: Missing values are legal, they must not throw.

File: nuxeo-runtime/nuxeo-runtime-kv/src/test/java/org/nuxeo/runtime/kv/AbstractKeyValueStoreTest.java [hunk 5]
```diff
@@ -115,6 +120,47 @@ public void testPutGet() {
         assertEquals(Collections.emptySet(), storeKeys());
     }
 
+    /**
+     * @since 9.10
+     */
+    protected void assertMapDeepEquals(Map map1, Map map2) {
+        assertTrue(String.format("Map1:%s and Map2:%s are not deepEquals", map1, map2),
+                new FastHashMap(map1).equals(new FastHashMap(map2)));
```
Comment (efge) [inline nuxeo-runtime/nuxeo-runtime-kv/src/test/java/org/nuxeo/runtime/kv/AbstractKeyValueStoreTest.java]: I don't get what `FastHashMap` brings over `HashMap`. Also, map comparison can be just done with `assertEquals(map1, map2)`

File: nuxeo-runtime/nuxeo-runtime-kv/src/test/java/org/nuxeo/runtime/kv/AbstractKeyValueStoreTest.java [hunk 6]
```diff
@@ -115,6 +120,47 @@ public void testPutGet() {
...
+     * @since 9.10
+     */
+    @Test
+    public void testGetMany() {
+        String key1 = "foo1";
+        String key2 = "foo2";
+        String key3 = "foo3";
+        Set<String> keys = new HashSet<>(Arrays.asList(key1, key2, key3));
+
+        assertMapDeepEquals(new HashMap(0), store.get(keys));
```
Comment (efge) [inline nuxeo-runtime/nuxeo-runtime-kv/src/test/java/org/nuxeo/runtime/kv/AbstractKeyValueStoreTest.java]: `new HashMap(0)` -> `Collections.emptyMap()`

File: nuxeo-core/nuxeo-core-redis/src/main/java/org/nuxeo/ecm/core/redis/contribs/RedisKeyValueStore.java [hunk 7]
```diff
@@ -128,6 +131,59 @@ public void put(String key, byte[] value, long ttl) {
         return redisExecutor.execute(jedis -> jedis.get(getBytes(namespace + key)));
     }
 
+    /**
+     * @since 9.10
+     */
+    @Override
+    public Map<String, byte[]> get(Set<String> keys) {
+        Objects.requireNonNull(keys);
+        Map<String, byte[]> kvMap = new HashMap<>(keys.size());
```
Comment (efge) [inline nuxeo-core/nuxeo-core-redis/src/main/java/org/nuxeo/ecm/core/redis/contribs/RedisKeyValueStore.java]: Could you rename the `kvMap` in this and other files too?

Comment (ron1) [reply nuxeo-core/nuxeo-core-redis/src/main/java/org/nuxeo/ecm/core/redis/contribs/RedisKeyValueStore.java]: The patch has been revised to also include these changes.

**Comment by efge** [comment]
Committed through #1483 1483

**Pull request closed without merging**